File: Prueba_iOS_ObjC/Prueba_iOS_ObjC/ImageCacheHandler.m
//
//  ImageCacheHandler.m
//  Prueba_iOS_ObjC
//
//  Created by Humberto Cetina on 2/26/17.
//  Copyright © 2017 Humberto Cetina. All rights reserved.
//

#import "ImageCacheHandler.h"


@implementation ImageCacheHandler

- (NSString *) imageNameFromUrl:(NSString *)url{
    
    NSArray *urlSeparated = [url componentsSeparatedByString:@"/"];
    NSString *imageName = urlSeparated.lastObject;
    return imageName;
}

- (void) imageForUrl:(NSString *)url withBlock:(void(^)(UIImage *image))block{
    
    NSURL *tmpFolderUrl = [NSURL fileURLWithPath:NSTemporaryDirectory() isDirectory:YES];
    
    NSString *imageName = [self imageNameFromUrl:url];
    NSURL *imageUrl = [tmpFolderUrl URLByAppendingPathComponent:imageName];
    UIImage *cacheImage = [UIImage imageWithContentsOfFile:imageUrl.path];
    
    if (cacheImage)
    {
        block(cacheImage);
    }
    else
    {
        NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
        NSCharacterSet *set = [NSCharacterSet URLQueryAllowedCharacterSet];
        NSString *imageNameEncoded = [url stringByAddingPercentEncodingWithAllowedCharacters:set];
        
        [[session dataTaskWithURL:[NSURL URLWithString:imageNameEncoded]
                completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
        {
            
            if (error)
            {
                block(nil);
            }
            else
            {
                [data writeToURL:imageUrl atomically:YES];
                UIImage *imageCached = [UIImage imageWithData:data];
                block(imageCached);
            }
        }] resume];
    }
}

@end

File: Prueba_iOS_Swift/Prueba_iOS/ImageCacheHandler.swift
//
//  ImageCacheHandler.swift
//  Prueba_iOS
//
//  Created by Humberto Cetina on 2/27/17.
//  Copyright © 2017 Humberto Cetina. All rights reserved.
//

import UIKit

class ImageCacheHandler: NSObject {

    func imageNameFrom(_ url: String?) -> String?{
        
        if let url = url
        {
            let urlSeparated: Array<String> = url.components(separatedBy: "/")
            return urlSeparated.last;
        }
        
        return nil
    }
    
    func imageForUrl(_ url: String?, andReturn block:@escaping (_ image: UIImage?) -> Void) {
        
        let tmpFolderUrl: URL = URL.init(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
        
        if let imageName = self.imageNameFrom(url)
        {
            let imageUrl: URL = tmpFolderUrl.appendingPathComponent(imageName)
            
            if let cacheImage = UIImage(contentsOfFile: imageUrl.path)
            {
                block(cacheImage); return
            }
                
            let sesion: URLSession = URLSession(configuration: URLSessionConfiguration.default)
            let set: CharacterSet = CharacterSet.urlQueryAllowed
            let imageNameEncoded: String = url!.addingPercentEncoding(withAllowedCharacters: set)!
            let imageServerUrl: URL? = URL(string: imageNameEncoded)
            
            if let imageServerUrl = imageServerUrl
            {
                let task: URLSessionDataTask = sesion.dataTask(with: imageServerUrl) { (data, urlResponse, error) in
                    
                    if error != nil
                    {
                        block(nil); return
                    }
                        
                    try! data!.write(to: imageUrl)
                    let imageToCache = UIImage(data: data!)
                    block(imageToCache)
                };
                
                task.resume();
            }
            else
            {
                block(nil)
            }
        }
        else
        {
            block(nil)
        }
    }
}
